### ISSUE TITLE
Athena types/functions with shared names

### ISSUE THREAD

**Issue by timcosta** [sqlfluff/sqlfluff#4749]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

In Athena/Trino/Hive, occasionally there are types and functions with the same name. An example of this is the `MAP<primitive, datatype>` type and the `MAP([keys...], [values...])` function. When casting to a map, the type in the cast should use angle braces so that it's a type, however the query engine will still work if you use parenthesis for whatever reason.

```sql
-- works, correct
SELECT CAST(JSON_PARSE('{"key": "value", "foo": "bar"}') AS MAP<varchar, varchar>)
-- works, incorrect
SELECT CAST(JSON_PARSE('{"key": "value", "foo": "bar"}') AS MAP(varchar, varchar))
```

It would be nice to be able to lint the second one so people change to the proper brackets, however there isnt a great way to do that currently.

The Map _type_ is defined as valid when there are angle braces, and not parens, so the second (incorrect but working) example above results in a `PRS` error due to sqlfluff expecting a type after `AS` not a function, and that prevents any linting from being done around the issue.

It doesnt feel great to me to add a second always-invalid map type declaration for parens just to lint the issue, but im not sure how else to do this.

Are there any outstanding patterns for linting something that is a PRS error, and shouldnt be considered valid sql?

### Dialect

athena

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by peterbud** [comment]
It is actually a parser error, not an enhancement.
